ai_powered_qa/components/agent.py

```diff
@@ -153,23 +153,26 @@ def _get_messages_for_completion(
         if user_prompt:
             total_tokens += count_tokens(user_prompt, model)
 
-        for i in range(len(self.history)):
+        messages_to_add = []
+        i = 0
+
+        while i < len(self.history):
```

Comment:
When we encouter a tool response in the history, we need to treat it as one unit with the corresponing tool_call message. We either need to include them both (or more, in case of multiple tool_calls), or nothing at all. That's why we need to skip iterations in this outer loop.